docs/migrations/README.md

```diff
@@ -0,0 +1,339 @@
+# Update to Cosmos-SDK v0.40
+
+The following document describes the changes, and an approach to update the chain to use Cosmos-SDK v0.40 
+aka stargate release.
+
+## Dependencies
+- Go 1.15+
+- Protoc 3.13.0
+- Node.js v12.0+ (Optional, for generating docs)
+
+A list of handy `make` commands are configured here: https://github.com/cosmos/cosmos-sdk/blob/v0.40.0-rc3/Makefile#L355-L443
+- `proto-update-deps` - To download/update the required thirdparty `proto` definitions
+- `proto-gen` - To auto generate proto code
+- `proto-check-breaking` - To check proto breaking changes
+- `proto-gen-any` - To generate the SDK's custom wrapper for google.protobuf.Any. It should only be run manually when needed
+- `proto-format` - To format proto files
+
+## Contents
+- [Updating a module to use Cosmos-SDK v0.40](#updating-a-module-to-use-cosmos-sdk-v040)
+- [Updating a cosmos-sdk chain to use SDK v0.40](#updating-a-cosmos-sdk-chain-to-use-sdk-v040)
+- [Upgrading a live chain to v0.40](#upgrading-a-live-chain-to-v040)
+
+## Updating a module to use Cosmos-SDK v0.40
+This section covers the changes in modules from `v0.39.x` to `v0.40`.
+### State
+- `internal` package is removed and `types`, `keeper` are moved to module level
+- `alias` usage is removed [#6311](https://github.com/cosmos/cosmos-sdk/issues/6311)
+#### types/codec.go
+- `codec.New()` is changed to `codec.NewLegacyAmino()`.
+- Added `RegisterInterfaces` method in which we add `RegisterImplementations` and `RegisterInterfaces` based on msgs and interfaces in module.
+
+    ```go
+    func RegisterInterfaces(registry types.InterfaceRegistry) {
+        registry.RegisterImplementations((*sdk.Msg)(nil),
+            &MsgSend{},
+            &MsgMultiSend{},
+        )
+
+        registry.RegisterInterface(
+            "cosmos.bank.v1beta1.SupplyI",
+            (*exported.SupplyI)(nil),
+            &Supply{},
+        )
+    }
+    ```
+- `init()` changed from:
+    ```go
+    func init() {
+	    RegisterCodec(cdc)
+    }
+    ```
+    to:
+    ```go
+    func init() {
+	    RegisterLegacyAminoCodec(amino)
+	    cryptocodec.RegisterCrypto(amino)
+    }
+    ```
+#### Msgs
+SDK now leverages protobuf service definitions for defining Msgs which will give us significant developer UX 
+improvements in terms of the code that is generated and the fact that return types will now be well defined.
+`sdk.Msg`'s are now replaced by protobuf services. Now every sdk.Msg is defined as a protobuf service method.
+Example: 
+```proto
+package cosmos.bank;
+
+service Msg {
+  rpc Send(MsgSend) returns (MsgSendResponse);
```

Comment:
should we add something about request type naming convention? (see 3rd paragraph of https://github.com/cosmos/cosmos-sdk/blob/1cc8af8d9be1cd85f628c8dd641d6d191098637a/docs/building-modules/messages-and-queries.md#msg-services)